File: core/static/core/js/vk-auth.js
/**
 * VK ID Widget integration (SDK 3.0+)
 * Handles VK authentication using new OAuth flow with code exchange
 */

(function() {
    'use strict';

    // Получаем конфигурацию из Django settings через data-атрибуты
    const getVkConfig = function() {
        const container = document.getElementById('vk-login-container');
        if (!container) return null;
        
        const appId = container.getAttribute('data-vk-app-id');
        const redirectUrl = container.getAttribute('data-vk-redirect-url') || window.location.origin + '/api/auth/vk/';
        
        if (!appId) {
            console.error('VK App ID не настроен в Django settings');
            return null;
        }
        
        return {
            appId: appId,
            redirectUrl: redirectUrl
        };
    };

    // Инициализация виджета VK ID
    const initVKWidget = function() {
        const container = document.getElementById('vk-login-container');
        if (!container) return;

        // Получаем конфигурацию
        const vkConfig = getVkConfig();
        if (!vkConfig) {
            createErrorButton(container, 'VK авторизация не настроена');
            return;
        }

        // Очищаем контейнер
        container.innerHTML = '';

        // Создаем скрипт VK SDK
        const script = document.createElement('script');
        script.async = true;
        script.src = 'https://unpkg.com/@vkid/sdk@3/dist-sdk/umd/index.js';
        
        script.onload = function() {
            if (!('VKIDSDK' in window)) {
                console.error('VK SDK не загрузился');
                createErrorButton(container, 'Не удалось загрузить VK SDK');
                return;
            }

            const VKID = window.VKIDSDK;

            try {
                // Инициализируем конфиг
                VKID.Config.init({
                    app: vkConfig.appId,
                    redirectUrl: vkConfig.redirectUrl,
                    responseMode: VKID.ConfigResponseMode.Callback,
                    source: VKID.ConfigSource.LOWCODE,
                    scope: '', // Можно добавить нужные права доступа
                });

                // Создаем экземпляр OneTap
                const oneTap = new VKID.OneTap();

                // Рендерим виджет
                oneTap.render({
                    container: container,
                    showAlternativeLogin: true
                })
                .on(VKID.WidgetEvents.ERROR, vkidOnError)
                .on(VKID.OneTapInternalEvents.LOGIN_SUCCESS, function (payload) {
                    const code = payload.code;
                    const deviceId = payload.device_id;

                    // Обмениваем code на токен и получаем данные пользователя
                    VKID.Auth.exchangeCode(code, deviceId)
                        .then(function(data) {
                            vkidOnSuccess(data);
                        })
                        .catch(vkidOnError);
                });

            } catch (e) {
                console.error('Ошибка инициализации VK ID:', e);
                createErrorButton(container, 'Ошибка инициализации VK ID');
            }
        };

        script.onerror = function() {
            console.error('Не удалось загрузить VK SDK');
            createErrorButton(container, 'Не удалось загрузить VK SDK');
        };

        container.appendChild(script);
    };

    // Обработка успешной авторизации
    const vkidOnSuccess = function(data) {
        // Данные содержат access_token и user info
        // Отправляем access_token на сервер для получения данных пользователя
        fetch('/api/auth/vk/', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'X-CSRFToken': getCookie('csrftoken')
            },
            body: JSON.stringify({
                'access_token': data.access_token,
                'user_id': data.user_id,
                'email': data.email || '',
                'first_name': data.first_name || '',
                'last_name': data.last_name || '',
                'photo_url': data.photo_url || ''
            })
        })
        .then(response => response.json())
        .then(data => {
            if (data.success) {
                // Перенаправляем пользователя в личный кабинет
                window.location.href = data.redirect_url || '/profile/';
            } else {
                alert('Ошибка авторизации: ' + (data.error || 'Неизвестная ошибка'));
            }
        })
        .catch(error => {
            console.error('Error:', error);
            alert('Произошла ошибка при авторизации. Попробуйте ещё раз.');
        });
    };

    // Обработка ошибки
    const vkidOnError = function(error) {
        console.error('VK ID Error:', error);
        alert('Ошибка авторизации ВКонтакте. Попробуйте ещё раз.');
    };

    // Кнопка с ошибкой если конфигурация не настроена или SDK не загрузился
    const createErrorButton = function(container, errorMessage) {
        container.innerHTML = `
            <div style="display: inline-block; padding: 10px 20px; background: #f44336; color: white; 
                         border-radius: 8px; font-weight: 500; cursor: not-allowed;">
                ${errorMessage}
            </div>
        `;
    };

    // Функция для получения CSRF токена
    const getCookie = function(name) {
        let cookieValue = null;
        if (document.cookie && document.cookie !== '') {
            const cookies = document.cookie.split(';');
            for (let i = 0; i < cookies.length; i++) {
                const cookie = cookies[i].trim();
                if (cookie.substring(0, name.length + 1) === (name + '=')) {
                    cookieValue = decodeURIComponent(cookie.substring(name.length + 1));
                    break;
                }
            }
        }
        return cookieValue;
    };

    // Экспорт функций для использования в других скриптах
    window.VKAuthUtils = {
        initVKWidget: initVKWidget,
        getCookie: getCookie
    };

    // Автоматическая инициализация при загрузке DOM
    document.addEventListener('DOMContentLoaded', function() {
        initVKWidget();
    });
})();

File: core/vk_auth.py
"""
Модуль для аутентификации через ВКонтакте (VK ID SDK 3.0+)

Использует новый OAuth flow с exchange code для получения данных пользователя.
Документация: https://dev.vk.com/vkid/
"""

import requests
from typing import Dict, Optional, Any
from django.conf import settings


def get_user_data_from_vk(access_token: str, user_id: int) -> Optional[Dict[str, Any]]:
    """
    Получает данные пользователя через VK API используя access_token
    
    Args:
        access_token: Токен доступа от VK ID
        user_id: ID пользователя ВКонтакте
        
    Returns:
        Словарь с данными пользователя или None если произошла ошибка
    """
    if not access_token:
        return None
    
    # VK API метод для получения информации о пользователе
    api_url = 'https://api.vk.com/method/users.get'
    params = {
        'user_ids': user_id,
        'fields': 'photo_200,first_name,last_name',
        'access_token': access_token,
        'v': '5.131'  # Версия API
    }
    
    try:
        response = requests.get(api_url, params=params, timeout=10)
        response.raise_for_status()
        data = response.json()
        
        if 'error' in data:
            print(f"VK API Error: {data['error']}")
            return None
        
        if 'response' in data and len(data['response']) > 0:
            user_data = data['response'][0]
            return {
                'vk_id': int(user_data['id']),
                'first_name': user_data.get('first_name', ''),
                'last_name': user_data.get('last_name', ''),
                'photo_url': user_data.get('photo_200', ''),
            }
        else:
            return None
            
    except requests.RequestException as e:
        print(f"Error calling VK API: {e}")
        return None


def validate_vk_oauth_data(access_token: str, user_id: int, client_secret: str) -> Optional[Dict[str, Any]]:
    """
    Проверяет данные авторизации от VK ID SDK 3.0+
    
    Args:
        access_token: Токен доступа от VK
        user_id: ID пользователя
        client_secret: Сервисный ключ доступа приложения
        
    Returns:
        Словарь с проверенными данными пользователя или None если проверка не пройдена
    """
    if not access_token or not user_id or not client_secret:
        return None
    
    # Получаем данные пользователя через VK API
    user_data = get_user_data_from_vk(access_token, user_id)
    
    if not user_data:
        return None
    
    # Дополнительно можно проверить токен через метод secure.checkToken
    # но для базовой авторизации достаточно успешного получения данных пользователя
    
    return user_data
